hicommonwealth/commonwealth
# Add Marlin Governance Integration (WIP)

## Description

- Integrate 2 Marlin Contracts, COMP/LIN Pond Contract + GovernorAlpha
- Create a Marlin chain adapter
  - Api
  - Governance controller, holder/holders/delegates. 
  - Proposals
- Views
  - [x] Proposal Page
  - [x] Create Proposal Page/Flow
  - [x] Delegation Page
  - [ ] Leaderboard page (needs The Graph or api to query top holders/accounts)
- [ ] Ensure that login and viewing proposals works with Metamask
- [ ] Ensure that login and viewing proposals works without Metamask

## Motivation and Context


Grow and integrate 🎉 

## Testing



The biggest thing to note is that you need to link the `@commonwealth/chain-events`, specifically the `zak.compound` branch prior to uncommenting the chain-event imports on the onchain pages and in the server (they're easy to find, marked with TODOs). 
`~/chain-events$ git checkout zak.marlin`, `yarn build`, `yarn link`
`cd ~/commonwealth`
`~/commonwealth$ yarn link @commonwealth/chain-events`

- Migration! Up and down should work as expected. `npx sequelize db:migrate`
- With a new account, sign in to marlin and link an address using metamask. 
  - Make sure metamask is set to ropsten or the chain will not load properly.
- Marlin Offchain features should work, discussions/threads/comments/etc. 
- Selecting any of the onchain-navigation items should cause the chain to load prior to rendering the page selected
- Selecting any of the other onchain-navigation items should render much quicker as chain/api has been turned on.
- Proposals page should be empty without adding any on a local-ganache. 
  - A bit of data about the user and contracts should appear in a 'stats'-like box at the top of the page. 
- Delegation of COMP/LIN should be done on the delegation page, which allows the user to paste the address they'd like to delegate to.
  - It should also should who you are currently delegated to.
- Create Proposal Page uses a form, though is a bit unwieldily as the value/calldata/signatures from contracts that are owned by GovernorAlpha are not necessarily human-legible. Pre-formatted data is required for ease of use.

## Have proper tags been added (for bug, enhancement, breaking change)?

## Does this PR affect any server routes?
